lib/storage/src/rbac/mod.rs

```diff
@@ -1,9 +1,209 @@
+        &self,
+        collection_name: &'a str,
+        whole_collection: bool,
+        mode: CollectionAccessMode,
+    ) -> Result<CollectionPass<'a>, StorageError> {
+        match self {
+            Access::Cluster(ClusterAccessMode::Read) => {
+                if mode != CollectionAccessMode::Read {
+                    return Err(StorageError::unauthorized(
+                        "Only read-only access is allowed",
+                    ));
+                }
+            }
+            Access::Cluster(ClusterAccessMode::ReadWrite) => (),
+            Access::Collection(rules) => {
+                let view = rules.find_view(collection_name)?;
+                if whole_collection {
+                    view.check_whole_access()?;
+                }
+                view.check_access_mode(mode)?;
+            }
+        }
+        Ok(CollectionPass(Cow::Borrowed(collection_name)))
+    }
+}
+
+impl AccessCollection {
+    pub(self) fn find_view<'a>(
+        &'a self,
+        collection_name: &'a str,
+    ) -> Result<AccessCollectionView<'a>, StorageError> {
+        let rule = self
+            .0
+            .iter()
+            .find(|collections| {
+                collections
+                    .collections
+                    .iter()
+                    .any(|name| name == collection_name)
+            })
+            .ok_or_else(|| {
+                StorageError::unauthorized(format!("Collection {collection_name} is not allowed"))
+            })?;
+        Ok(AccessCollectionView {
+            collection: collection_name,
+            access: rule.access,
+            payload: &rule.payload,
+        })
+    }
+}
+
+struct AccessCollectionView<'a> {
+    pub collection: &'a str,
+    pub access: CollectionAccessMode,
+    pub payload: &'a Option<PayloadConstraint>,
+}
+
+impl<'a> AccessCollectionView<'a> {
+    pub(self) fn check_whole_access(&self) -> Result<(), StorageError> {
+        if self.payload.is_some() {
+            return incompatible_with_payload_constraint(self.collection);
+        }
+        Ok(())
+    }
+
+    pub(self) fn check_access_mode(
+        &self,
+        min_access: CollectionAccessMode,
+    ) -> Result<(), StorageError> {
+        if self.access < min_access {
+            return Err(StorageError::unauthorized(format!(
+                "Collection {} is not allowed",
+                self.collection,
+            )));
+        }
+        Ok(())
+    }
+}
+
+pub struct CollectionMultipass;
+
+impl CollectionMultipass {
+    pub fn issue_pass<'a>(&self, name: &'a str) -> CollectionPass<'a> {
+        CollectionPass(Cow::Borrowed(name))
+    }
+}
+
+/// A pass that allows access to a specific collection.
+pub struct CollectionPass<'a>(pub(self) Cow<'a, str>);
+
+impl<'a> CollectionPass<'a> {
+    pub fn name(&'a self) -> &'a str {
+        &self.0
+    }
+
+    pub fn into_static(self) -> CollectionPass<'static> {
+        CollectionPass(Cow::Owned(self.0.into_owned()))
+    }
+}
```

Comment:
Doesn't implementing [`to_owned`](https://doc.rust-lang.org/std/borrow/trait.ToOwned.html) instead make more sense?

Reply:
This would require to have two distinct types (e.g. `CollectionPassOwned` and `CollectionPassRef<'a>`). Having a single type with a `Cow` inside is more elegant, IMO.

Reply:
Having a :cow2: is fine. Though nitpicky, maybe we can rename the function to `to_owned()`  then to follow naming conventions.